### ISSUE TITLE
qdevice does not work if ring0 is not configured

### ISSUE THREAD

**Issue by Fabian-Gruenbichler** [corosync/corosync-qdevice#16]
the following debug log gets produced on the qdevice side:
```
# /usr/sbin/corosync-qdevice -f -d
Nov 11 20:39:45 debug   Initializing votequorum
Nov 11 20:39:45 debug   Initializing local socket
Nov 11 20:39:45 debug   Registering qdevice models
Nov 11 20:39:45 debug   Configuring qdevice
Nov 11 20:39:45 debug   Configuring master_wins
Nov 11 20:39:45 debug   Getting configuration node list
Nov 11 20:39:45 debug   Initializing qdevice model
Nov 11 20:39:45 debug   Initializing qdevice_net_instance
Nov 11 20:39:45 debug   Registering algorithms
Nov 11 20:39:45 debug   Initializing NSS
Nov 11 20:39:45 debug   Cast vote timer remains stopped.
Nov 11 20:39:45 debug   Initializing cmap tracking
Nov 11 20:39:45 debug   Waiting for ring id
Nov 11 20:39:45 debug   Votequorum nodelist notify callback:
Nov 11 20:39:45 debug     Ring_id = (1.12)
Nov 11 20:39:45 debug     Node list (size = 2):
Nov 11 20:39:45 debug       0 nodeid = 1
Nov 11 20:39:45 debug       1 nodeid = 2
Nov 11 20:39:45 debug   Algorithm decided to pause cast vote timer and result vote is No change
Nov 11 20:39:45 debug   Cast vote timer is now paused.
Nov 11 20:39:45 debug   Votequorum quorum notify callback:
Nov 11 20:39:45 debug     Quorate = 1
Nov 11 20:39:45 debug     Node list (size = 3):
Nov 11 20:39:45 debug       0 nodeid = 1, state = 1
Nov 11 20:39:45 debug       1 nodeid = 2, state = 1
Nov 11 20:39:45 debug       2 nodeid = 0, state = 0
Nov 11 20:39:45 debug   Algorithm decided to not send list and result vote is No change
Nov 11 20:39:45 debug   Waiting for initial heuristics exec result
Nov 11 20:39:45 debug   worker: Heuristic worker initialized
Nov 11 20:39:45 debug   worker: qdevice_heuristics_worker_cmd_process_one_line: Received exec-list-clear command
Nov 11 20:39:45 debug   worker: qdevice_heuristics_worker_cmd_process_exec: Received exec command with seq_no "1" and timeout "15000"
Nov 11 20:39:45 debug   Received heuristics exec result command with seq_no "1" and result "Disabled"
Nov 11 20:39:45 debug   Votequorum heuristics exec result callback:
Nov 11 20:39:45 debug     seq_number = 1, exec_result = Disabled
Nov 11 20:39:45 debug   Algorithm decided to not send list, result vote is No change and heuristics is Undefined
Nov 11 20:39:45 debug   Cast vote timer is no longer paused.
Nov 11 20:39:45 debug   Not scheduling heuristics timer because mode is not enabled
Nov 11 20:39:45 debug   Running qdevice model
Nov 11 20:39:45 debug   Executing qdevice-net
Nov 11 20:39:45 debug   Trying connect to qnetd server IP1:5403 (timeout = 8000ms)
Nov 11 20:39:45 debug   Sending preinit msg to qnetd
Nov 11 20:39:45 debug   Received preinit reply msg
Nov 11 20:39:45 debug   Sending client auth data.
Nov 11 20:39:45 debug   Received init reply msg
Nov 11 20:39:45 debug   Scheduling send of heartbeat every 8000ms
Nov 11 20:39:45 debug   Executing after-connect heuristics.
Nov 11 20:39:45 debug   worker: qdevice_heuristics_worker_cmd_process_exec: Received exec command with seq_no "2" and timeout "5000"
Nov 11 20:39:45 debug   Received heuristics exec result command with seq_no "2" and result "Disabled"
Nov 11 20:39:45 debug   Algorithm decided to send config node list, send membership node list, send quorum node list, heuristics is Undefined and result vote is Wait for reply
Nov 11 20:39:45 debug   Sending config node list seq = 4
Nov 11 20:39:45 debug     Node list:
Nov 11 20:39:45 debug   Sending membership node list seq = 5, ringid = (1.12), heuristics = Undefined.
Nov 11 20:39:45 debug     Node list:
Nov 11 20:39:45 debug       0 node_id = 1, data_center_id = 0, node_state = not set
Nov 11 20:39:45 debug       1 node_id = 2, data_center_id = 0, node_state = not set
Nov 11 20:39:45 debug   Sending quorum node list seq = 6, quorate = 1
Nov 11 20:39:45 debug     Node list:
Nov 11 20:39:45 debug       0 node_id = 1, data_center_id = 0, node_state = member
Nov 11 20:39:45 debug       1 node_id = 2, data_center_id = 0, node_state = member
Nov 11 20:39:45 debug   Cast vote timer remains stopped.
Nov 11 20:39:45 error   Received server error 18. Disconnecting from server
Nov 11 20:39:45 debug   Algorithm result vote is NACK
Nov 11 20:39:45 debug   Cast vote timer is now scheduled every 5000ms voting NACK.
Nov 11 20:39:45 debug   Removing cmap tracking
Nov 11 20:39:45 debug   Destroying qdevice model
Nov 11 20:39:45 debug   Destroying algorithm
Nov 11 20:39:45 debug   Destroying qdevice_net_instance
Nov 11 20:39:45 debug   Shutting down NSS
Nov 11 20:39:45 debug   Destroying qdevice ipc
Nov 11 20:39:45 debug   Destroying votequorum and cmap
Nov 11 20:39:45 debug   Destroying heuristics
Nov 11 20:39:45 debug   Waiting for heuristics worker to finish
Nov 11 20:39:45 debug   worker: Heuristic worker shutdown requested
Nov 11 20:39:45 debug   worker: Waiting for all processes to exit
Nov 11 20:39:45 debug   Closing log
```

the following on the qnet side:
```
# corosync-qnetd -f -d
Nov 12 12:29:48 debug   Initializing nss
Nov 12 12:29:48 debug   Initializing local socket
Nov 12 12:29:48 debug   Creating listening socket
Nov 12 12:29:48 debug   Registering algorithms
Nov 12 12:29:48 debug   QNetd ready to provide service
Nov 12 12:30:01 debug   New client connected
Nov 12 12:30:01 debug     cluster name = modal
Nov 12 12:30:01 debug     tls started = 1
Nov 12 12:30:01 debug     tls peer certificate verified = 1
Nov 12 12:30:01 debug     node_id = 1
Nov 12 12:30:01 debug     pointer = 0x18576c8
Nov 12 12:30:01 debug     addr_str = ::ffff:IP2:44332
Nov 12 12:30:01 debug     ring id = (1.12)
Nov 12 12:30:01 debug     cluster dump:
Nov 12 12:30:01 debug       client = ::ffff:IP2:44332, node_id = 1
Nov 12 12:30:01 debug   Client ::ffff:IP2:44332 (cluster modal, node_id 1) sent initial node list.
Nov 12 12:30:01 debug     msg seq num = 4
Nov 12 12:30:01 debug     config version = 11
Nov 12 12:30:01 debug     node list:
Nov 12 12:30:01 error   ffsplit: Received empty config node list for client ::ffff:IP2:44332
Nov 12 12:30:01 error   Algorithm returned error code. Sending error reply.
Nov 12 12:30:01 debug   Client ::ffff:IP2:44332 (cluster modal, node_id 1) sent membership node list.
Nov 12 12:30:01 debug     msg seq num = 5
Nov 12 12:30:01 debug     ring id = (1.12)
Nov 12 12:30:01 debug     heuristics = Undefined
Nov 12 12:30:01 debug     node list:
Nov 12 12:30:01 debug       node_id = 1, data_center_id = 0, node_state = not set
Nov 12 12:30:01 debug       node_id = 2, data_center_id = 0, node_state = not set
Nov 12 12:30:01 debug   Algorithm result vote is Ask later
Nov 12 12:30:01 debug   Client ::ffff:IP2:44332 (cluster modal, node_id 1) sent quorum node list.
Nov 12 12:30:01 debug     msg seq num = 6
Nov 12 12:30:01 debug     quorate = 1
Nov 12 12:30:01 debug     node list:
Nov 12 12:30:01 debug       node_id = 2, data_center_id = 0, node_state = member
Nov 12 12:30:01 debug       node_id = 1, data_center_id = 0, node_state = member
Nov 12 12:30:01 debug   Algorithm result vote is No change
Nov 12 12:30:01 debug   Client closed connection
Nov 12 12:30:01 debug   Client ::ffff:IP2:44332 (init_received 1, cluster modal, node_id 1) disconnect
Nov 12 12:30:01 debug   ffsplit: Membership for cluster modal is now stable
Nov 12 12:30:01 debug   ffsplit: No quorate partition was selected
Nov 12 12:30:01 debug   ffsplit: No client gets NACK
Nov 12 12:30:01 debug   ffsplit: No client gets ACK
Nov 12 12:30:01 debug   New client connected
Nov 12 12:30:01 debug     cluster name = modal
Nov 12 12:30:01 debug     tls started = 1
Nov 12 12:30:01 debug     tls peer certificate verified = 1
Nov 12 12:30:01 debug     node_id = 1
Nov 12 12:30:01 debug     pointer = 0x18576c8
Nov 12 12:30:01 debug     addr_str = ::ffff:IP2:44348
Nov 12 12:30:01 debug     ring id = (1.12)
Nov 12 12:30:01 debug     cluster dump:
Nov 12 12:30:01 debug       client = ::ffff:IP2:44348, node_id = 1
Nov 12 12:30:01 debug   Client ::ffff:IP2:44348 (cluster modal, node_id 1) sent initial node list.
Nov 12 12:30:01 debug     msg seq num = 4
Nov 12 12:30:01 debug     config version = 11
Nov 12 12:30:01 debug     node list:
Nov 12 12:30:01 error   ffsplit: Received empty config node list for client ::ffff:IP2:44348
Nov 12 12:30:01 error   Algorithm returned error code. Sending error reply.
Nov 12 12:30:01 debug   Client ::ffff:IP2:44348 (cluster modal, node_id 1) sent membership node list.
Nov 12 12:30:01 debug     msg seq num = 5
Nov 12 12:30:01 debug     ring id = (1.12)
Nov 12 12:30:01 debug     heuristics = Undefined
Nov 12 12:30:01 debug     node list:
Nov 12 12:30:01 debug       node_id = 1, data_center_id = 0, node_state = not set
Nov 12 12:30:01 debug       node_id = 2, data_center_id = 0, node_state = not set
Nov 12 12:30:01 debug   Algorithm result vote is Ask later
Nov 12 12:30:01 debug   Client ::ffff:IP2:44348 (cluster modal, node_id 1) sent quorum node list.
Nov 12 12:30:01 debug     msg seq num = 6
Nov 12 12:30:01 debug     quorate = 1
Nov 12 12:30:01 debug     node list:
Nov 12 12:30:01 debug       node_id = 2, data_center_id = 0, node_state = member
Nov 12 12:30:01 debug       node_id = 1, data_center_id = 0, node_state = member
Nov 12 12:30:01 debug   Algorithm result vote is No change
Nov 12 12:30:01 debug   Client closed connection
Nov 12 12:30:01 debug   Client ::ffff:IP2:44348 (init_received 1, cluster modal, node_id 1) disconnect
Nov 12 12:30:01 debug   ffsplit: Membership for cluster modal is now stable
Nov 12 12:30:01 debug   ffsplit: No quorate partition was selected
Nov 12 12:30:01 debug   ffsplit: No client gets NACK
Nov 12 12:30:01 debug   ffsplit: No client gets ACK
Nov 12 12:30:02 debug   New client connected
Nov 12 12:30:02 debug     cluster name = modal
Nov 12 12:30:02 debug     tls started = 1
Nov 12 12:30:02 debug     tls peer certificate verified = 1
Nov 12 12:30:02 debug     node_id = 1
Nov 12 12:30:02 debug     pointer = 0x18576c8
Nov 12 12:30:02 debug     addr_str = ::ffff:IP2:44354
Nov 12 12:30:02 debug     ring id = (1.12)
Nov 12 12:30:02 debug     cluster dump:
Nov 12 12:30:02 debug       client = ::ffff:IP2:44354, node_id = 1
Nov 12 12:30:02 debug   Client ::ffff:IP2:44354 (cluster modal, node_id 1) sent initial node list.
Nov 12 12:30:02 debug     msg seq num = 4
Nov 12 12:30:02 debug     config version = 11
Nov 12 12:30:02 debug     node list:
Nov 12 12:30:02 error   ffsplit: Received empty config node list for client ::ffff:IP2:44354
Nov 12 12:30:02 error   Algorithm returned error code. Sending error reply.
Nov 12 12:30:02 debug   Client ::ffff:IP2:44354 (cluster modal, node_id 1) sent membership node list.
Nov 12 12:30:02 debug     msg seq num = 5
Nov 12 12:30:02 debug     ring id = (1.12)
Nov 12 12:30:02 debug     heuristics = Undefined
Nov 12 12:30:02 debug     node list:
Nov 12 12:30:02 debug       node_id = 1, data_center_id = 0, node_state = not set
Nov 12 12:30:02 debug       node_id = 2, data_center_id = 0, node_state = not set
Nov 12 12:30:02 debug   Algorithm result vote is Ask later
Nov 12 12:30:02 debug   Client ::ffff:IP2:44354 (cluster modal, node_id 1) sent quorum node list.
Nov 12 12:30:02 debug     msg seq num = 6
Nov 12 12:30:02 debug     quorate = 1
Nov 12 12:30:02 debug     node list:
Nov 12 12:30:02 debug       node_id = 2, data_center_id = 0, node_state = member
Nov 12 12:30:02 debug       node_id = 1, data_center_id = 0, node_state = member
Nov 12 12:30:02 debug   Algorithm result vote is No change
Nov 12 12:30:02 debug   Client closed connection
Nov 12 12:30:02 debug   Client ::ffff:IP2:44354 (init_received 1, cluster modal, node_id 1) disconnect
Nov 12 12:30:02 debug   ffsplit: Membership for cluster modal is now stable
Nov 12 12:30:02 debug   ffsplit: No quorate partition was selected
Nov 12 12:30:02 debug   ffsplit: No client gets NACK
Nov 12 12:30:02 debug   ffsplit: No client gets ACK
Nov 12 12:30:02 debug   New client connected
Nov 12 12:30:02 debug     cluster name = modal
Nov 12 12:30:02 debug     tls started = 1
Nov 12 12:30:02 debug     tls peer certificate verified = 1
Nov 12 12:30:02 debug     node_id = 1
Nov 12 12:30:02 debug     pointer = 0x18576c8
Nov 12 12:30:02 debug     addr_str = ::ffff:IP2:44366
Nov 12 12:30:02 debug     ring id = (1.12)
Nov 12 12:30:02 debug     cluster dump:
Nov 12 12:30:02 debug       client = ::ffff:IP2:44366, node_id = 1
Nov 12 12:30:02 debug   Client ::ffff:IP2:44366 (cluster modal, node_id 1) sent initial node list.
Nov 12 12:30:02 debug     msg seq num = 4
Nov 12 12:30:02 debug     config version = 11
Nov 12 12:30:02 debug     node list:
Nov 12 12:30:02 error   ffsplit: Received empty config node list for client ::ffff:IP2:44366
Nov 12 12:30:02 error   Algorithm returned error code. Sending error reply.
Nov 12 12:30:02 debug   Client ::ffff:IP2:44366 (cluster modal, node_id 1) sent membership node list.
Nov 12 12:30:02 debug     msg seq num = 5
Nov 12 12:30:02 debug     ring id = (1.12)
Nov 12 12:30:02 debug     heuristics = Undefined
Nov 12 12:30:02 debug     node list:
Nov 12 12:30:02 debug       node_id = 1, data_center_id = 0, node_state = not set
Nov 12 12:30:02 debug       node_id = 2, data_center_id = 0, node_state = not set
Nov 12 12:30:02 debug   Algorithm result vote is Ask later
Nov 12 12:30:02 debug   Client ::ffff:IP2:44366 (cluster modal, node_id 1) sent quorum node list.
Nov 12 12:30:02 debug     msg seq num = 6
Nov 12 12:30:02 debug     quorate = 1
Nov 12 12:30:02 debug     node list:
Nov 12 12:30:02 debug       node_id = 2, data_center_id = 0, node_state = member
Nov 12 12:30:02 debug       node_id = 1, data_center_id = 0, node_state = member
Nov 12 12:30:02 debug   Algorithm result vote is No change
Nov 12 12:30:02 debug   Client closed connection
Nov 12 12:30:02 debug   Client ::ffff:IP2:44366 (init_received 1, cluster modal, node_id 1) disconnect
Nov 12 12:30:02 debug   ffsplit: Membership for cluster modal is now stable
Nov 12 12:30:02 debug   ffsplit: No quorate partition was selected
Nov 12 12:30:02 debug   ffsplit: No client gets NACK
Nov 12 12:30:02 debug   ffsplit: No client gets ACK
Nov 12 12:30:02 debug   New client connected
Nov 12 12:30:02 debug     cluster name = modal
Nov 12 12:30:02 debug     tls started = 1
Nov 12 12:30:02 debug     tls peer certificate verified = 1
Nov 12 12:30:02 debug     node_id = 1
Nov 12 12:30:02 debug     pointer = 0x18576c8
Nov 12 12:30:02 debug     addr_str = ::ffff:IP2:44370
Nov 12 12:30:02 debug     ring id = (1.12)
Nov 12 12:30:02 debug     cluster dump:
Nov 12 12:30:02 debug       client = ::ffff:IP2:44370, node_id = 1
Nov 12 12:30:02 debug   Client ::ffff:IP2:44370 (cluster modal, node_id 1) sent initial node list.
Nov 12 12:30:02 debug     msg seq num = 4
Nov 12 12:30:02 debug     config version = 11
Nov 12 12:30:02 debug     node list:
Nov 12 12:30:02 error   ffsplit: Received empty config node list for client ::ffff:IP2:44370
Nov 12 12:30:02 error   Algorithm returned error code. Sending error reply.
Nov 12 12:30:02 debug   Client ::ffff:IP2:44370 (cluster modal, node_id 1) sent membership node list.
Nov 12 12:30:02 debug     msg seq num = 5
Nov 12 12:30:02 debug     ring id = (1.12)
Nov 12 12:30:02 debug     heuristics = Undefined
Nov 12 12:30:02 debug     node list:
Nov 12 12:30:02 debug       node_id = 1, data_center_id = 0, node_state = not set
Nov 12 12:30:02 debug       node_id = 2, data_center_id = 0, node_state = not set
Nov 12 12:30:02 debug   Algorithm result vote is Ask later
Nov 12 12:30:02 debug   Client ::ffff:IP2:44370 (cluster modal, node_id 1) sent quorum node list.
Nov 12 12:30:02 debug     msg seq num = 6
Nov 12 12:30:02 debug     quorate = 1
Nov 12 12:30:02 debug     node list:
Nov 12 12:30:02 debug       node_id = 2, data_center_id = 0, node_state = member
Nov 12 12:30:02 debug       node_id = 1, data_center_id = 0, node_state = member
Nov 12 12:30:02 debug   Algorithm result vote is No change
Nov 12 12:30:02 debug   Client closed connection
Nov 12 12:30:02 debug   Client ::ffff:IP2:44370 (init_received 1, cluster modal, node_id 1) disconnect
Nov 12 12:30:02 debug   ffsplit: Membership for cluster modal is now stable
Nov 12 12:30:02 debug   ffsplit: No quorate partition was selected
Nov 12 12:30:02 debug   ffsplit: No client gets NACK
Nov 12 12:30:02 debug   ffsplit: No client gets ACK
```

with a corosync.conf with just ring/link 5 defined.

the following looks like a likely culprit;
https://github.com/corosync/corosync-qdevice/blob/master/qdevices/qdevice-cmap.c#L121

switching to ring0 instead of 5 makes qdevice/qnetd happy. possible we could continue on any non `ringX_addr` key, and check for a node list entry before adding to avoid adding a node multiple times if multiple rings are configured?

**Comment by jfriesse** [comment]
@Fabian-Gruenbichler Thank you for the report, really really nice catch!

I've tried to make a fix - PR #17 - so I would be more than grateful if you could give it a try.

**Comment by jfriesse** [comment]
I've merge PR #17 so problem should be fixed now so closing this issue.
